src/server/page_server.cpp

```diff
@@ -525,14 +542,41 @@ page_server::disconnect_all_tran_server ()
     {
       for (size_t i = 0; i < m_passive_tran_server_conn.size (); i++)
 	{
-	  er_log_debug (ARG_FILE_LINE, "disconnect_all_tran_server:"
-			" Disconnected passive transaction server with channel id: %s.\n",
+	  er_log_debug (ARG_FILE_LINE,
+			"disconnect_all_tran_server: Request passive transaction server to disconnect. Channel id: %s.\n",
 			m_passive_tran_server_conn[i]->get_connection_id ().c_str ());
 	  m_passive_tran_server_conn[i]->remove_prior_sender_sink ();
-	  m_passive_tran_server_conn[i].reset (nullptr);
+	  m_passive_tran_server_conn[i]->push_disconnection_request ();
 	}
-      m_passive_tran_server_conn.clear ();
     }
+
+  er_log_debug (ARG_FILE_LINE,
+		"disconnect_all_tran_server: Wait until all connections are disconnected.\n");
+
+  /*
+   *  m_active_tran_server_conn == nullptr : The ATS has been disconnected or disconncted in progress (by async_disconnect_handler)
+   *  The connection for a PTS is not in m_passive_tran_server_conn : the PTS has been disconnected or disconnected in progress(by async_disconnect_handler)
+   *
+   *  The disconnection of a TS starts when tran_to_page_request::SEND_DISCONNECT_MSG is arrived from the TS. It's either tiggered by page_to_tran_request::SEND_DISCONNECT_REQUEST_MSG from this page server or just initiated by the TS itself.
+   *
+   *  If some disconnections are underway, they will be waited for at the m_async_disconnect_handler.terminate () below.
+   */
+  {
+    auto no_conn_alive_pred = [this]
+    {
+      return m_active_tran_server_conn == nullptr && m_passive_tran_server_conn.empty();
+    };
+
+    std::unique_lock ulock { m_conn_mutex };
+    m_conn_cv.wait (ulock, no_conn_alive_pred);
```

Comment:
> I'm not against it, but I'd like to know the advantage of wait for instead of wait, which I asked before but not replied. Is it for a future extension like timeout? or, does it have another advantage itself?

No, no advantage other than being able to add a timeout afterwards.
In my opinion, I also think that `wait` is the more conservative to actually use but I saw that `wait_for` is used a lot in the codebase.
So, in truth, I do not have a good reason to push for `wait_for` other than being able to easily add a timeout in the future. But, for now, we don't need that timeout so - `wait` it is.

Reply:
@wrlawodms 

> You're right. Thank you for pointing out. Another way would be to stop css_master_thread() before this function, or prevent a new connection is created in the function with a flag at least. Do you want me to create the issue? I think it can be done after discussing new regression tests.

yes, separate issue

Reply:
added: http://jira.cubrid.org/browse/LETS-679